### PR TITLE
Suggestions

### PR DESCRIPTION
**Pull request by robertwilkinson-dev**
This PR:
- channel router refactor
- channel controller refactor
- adds validation middleware
- adds validation schema example
- updates env vars

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+PORT=3000
+DB_HOST=localhost
+DB_DB=slack
+DB_USERNAME=YOUR_USERNAME
+DB_PASSWORD=YOUR_PASSWORD
```
Comment (robertwilkinson-dev) [inline .env.example:5]: added an example as its easy to share env vars without showing secrets

File: custom.d.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+declare namespace NodeJS {
+  interface ProcessEnv {
+    NODE_ENV: 'development' | 'production' | 'test'
+    DB_HOST: string;
+    DB_DB: string;
+    DB_USERNAME: string;
+    DB_PASSWORD: string;
+  }
+}
```
Comment (robertwilkinson-dev) [inline custom.d.ts:9]: added environment typings for `process.env.XXX`

File: package.json [hunk 3]
```diff
@@ -11,6 +11,7 @@
   },
   "dependencies": {
     "express": "^4.18.2",
+    "joi": "^17.9.2",
```
Comment (robertwilkinson-dev) [inline package.json:14]: `joi` for validation schemas

File: src/controllers/channelController/addUserToChannel.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    if (!channel) {
+      throw new CustomError(`Channel with id ${channelId} not found`, 404);
+    }
+    await channel.addUser(userId);
+    res.json(channel);
+
+  } catch (error) {
+    throw new CustomError(`unable to add user ${userId} to channel ${channelId}`, 500);
+  }
+};
```
Comment (robertwilkinson-dev) [inline src/controllers/channelController/addUserToChannel.ts:20]: refactor channel controller [PART 1]

File: src/controllers/channelController/createChannel.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  try {
+    const channel = await Channel.create({
+      channelName,
+    });
+  
+    res.json(channel);
+  } catch (error) {
+    throw new CustomError(`unable to create channel ${channelName}`, 500);
+  }
+};
```
Comment (robertwilkinson-dev) [inline src/controllers/channelController/createChannel.ts:17]: refactor channel controller [PART 2]

File: src/controllers/channelController/getAllChannels.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+export const getAllChannels = async (req: Request, res: Response) => {
+  try {
+    const channels = await Channel.findAll();
+    res.json(channels);
+
+  } catch (error) {
+    throw new CustomError("Something went wrong", 500);
+  }
+};
+
```
Comment (robertwilkinson-dev) [inline src/controllers/channelController/getAllChannels.ts:14]: refactor channel controller [PART 3]

File: src/controllers/channelController/index.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+import { createChannel } from './createChannel';
+import { addUserToChannel } from './addUserToChannel';
+
+const channelController = {
+  getAllChannels,
+  createChannel,
+  addUserToChannel
+}
+
+export default channelController;
```
Comment (robertwilkinson-dev) [inline src/controllers/channelController/index.ts:11]: refactor channel controller [PART 4] here we barrel all the methods into one object called `channelController` this makes the code more readable later on in the router

File: src/controllers/usersController.ts [hunk 8]
```diff
@@ -16,7 +16,7 @@ export const createNewUser = async (req: Request, res: Response) => {
     },
   });
   if (user) {
-    throw new customError("User already exists", 400);
+    throw new CustomError("User already exists", 400);
   }
 
   const user1 = await User.create({
```
Comment (robertwilkinson-dev) [inline src/controllers/usersController.ts:22]: just updated the class to have a capital letter

File: src/index.ts [hunk 9]
```diff
@@ -22,7 +22,7 @@ app.use("/api/users", usersRoutes);
 app.use("/api/channels", channelRoutes);
 
 app.use("*", (req: Request, res: Response, next: NextFunction) => {
-  const error = new customError(`Route ${req.originalUrl} not found!!!`, 404);
+  const error = new CustomError(`Route ${req.originalUrl} not found!!!`, 404);
   next(error);
 });
```
Comment (robertwilkinson-dev) [inline src/index.ts:28]: your middleware was fine so I renamed the file and updated the imports

File: src/middlewares/globalErrorHandler.ts [hunk 10]
```diff
@@ -13,23 +17,17 @@ interface ErrorResponse {
...
-  err.statusCode = err.statusCode || 500;
-
-  const errorResponse: ErrorResponse = {
+    const errorResponse: ErrorResponse = {
     status: err.statusCode < 500 ? "fail" : "error",
     message: err.message,
   };
 
   res.status(err.statusCode).json(errorResponse);
 };
```
Comment (robertwilkinson-dev) [inline src/middlewares/globalErrorHandler.ts:33]: more import updates

File: src/middlewares/validateSchema.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  return (req: Request, res: Response, next: NextFunction) => {
+    const { error }: ValidationResult = schema.validate(req.body);
+
+    if (error) {
+      return next(new CustomError('ValidationError', 400));
+    }
+
+    next();
+  };
+};
```
Comment (robertwilkinson-dev) [inline src/middlewares/validateSchema.ts:17]: a basic validation middleware using `joi`

File: src/models/sequelize.ts [hunk 12]
```diff
@@ -2,9 +2,19 @@ import { Sequelize } from "sequelize";
...
+  throw new Error('Missing required environment variables for database connection.');
+}
+
+const sequelize = new Sequelize(DB_DB, DB_USERNAME, DB_PASSWORD, {
+  host: DB_HOST,
+  dialect: 'postgres',
+});
 
 const testDbConnection = async (sequelize: Sequelize) => {
   try {
```
Comment (robertwilkinson-dev) [inline src/models/sequelize.ts:20]: just updated the db connection according to the docs so it worked with docker (you can ignore this bit if you want)

File: src/routes/channelsRoutes.ts [hunk 13]
```diff
@@ -1,34 +1,12 @@
...
-    await addUserToChannel(req, res);
-  } catch (error) {
-    next(error);
-  }
-});
+router.get('/', channelController.getAllChannels);
+router.post('/', validateSchema(channelSchema), channelController.createChannel);
+router.post('/addUser', channelController.addUserToChannel);
 
 export default router;
```
Comment (robertwilkinson-dev) [inline src/routes/channelsRoutes.ts:12]: clean up router file:
- add validation
- update usage of controller

File: src/schemas/channel.schema.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+import * as Joi from 'joi';
+
+export const channelSchema = Joi.object({
+  channelName: Joi.string().required(),
+});
```
Comment (robertwilkinson-dev) [inline src/schemas/channel.schema.ts:5]: basic schema for validation example

File: tsconfig.json [hunk 15]
```diff
@@ -7,8 +7,8 @@
     "esModuleInterop": true,
     "forceConsistentCasingInFileNames": true,
     "strict": true,
-    "resolveJsonModule": true
+    "resolveJsonModule": true,
   },
-  "include": ["src/**/*.ts"],
+  "include": ["src/**/*.ts", "custom.d.ts"],
   "exclude": ["node_modules"]
 }
```
Comment (robertwilkinson-dev) [inline tsconfig.json:14]: update tsconfig to read environment typings

**Review by musayuksel — COMMENTED** [review]
Thank you for all your suggestions Rob

**Pull request merged**